### PR TITLE
Take in multiple stitches such as "k5"

### PR DESCRIPTION
**Pull request by graceolivia**
Added to method in NestedArrayBuilder that takes something like `"k5"` and converts it to `"k1 k1 k1 k1 k1" `. This allows internal transformation is done right before the MetaDataBuilder stage where the pattern gets turned into an ASCII chart.

### CODE REVIEW

File: Sources/SwiftCliCore/NestedArrayBuilder.swift [hunk 1]
```diff
@@ -8,4 +8,29 @@ public class NestedArrayBuilder {
         let rowStitches = substringRowStitches.map {(String($0))}
         return(rowStitches)
     }
+
+    public func expandRow(row: [String]) -> [String] {
+        let expandedRow = row.flatMap { getMultipleStitchInfo(stitch: $0) }
+        return expandedRow
```
Comment (breadoliveoilsalt) [inline Sources/SwiftCliCore/NestedArrayBuilder.swift:14]: Could be just `return row.flatMap...etc.`

File: Sources/SwiftCliCore/InputValidator.swift [hunk 2]
```diff
@@ -30,7 +30,9 @@ public class InputValidator {
             }
         }
 
-        let patternMetaData = MetaDataBuilder().buildAllMetaData(stitchArray: patternNestedArray )
+        let expandedNestedArray = patternNestedArray.map { NestedArrayBuilder().expandRow(row: $0) }
+
+        let patternMetaData = MetaDataBuilder().buildAllMetaData(stitchArray: expandedNestedArray)
```
Comment (hnlee) [inline Sources/SwiftCliCore/InputValidator.swift:35]: Consider refactoring this `inputValidation()` method for this PR. Suggestions:
- Use a verb name
- Split into the "three phases" you mentioned during pairing so that this method's responsibility is just to make sure the phases happen in the right order.

File: Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift [hunk 3]
```diff
@@ -16,3 +16,13 @@ class NestedArrayMakerTest: XCTestCase {
...
 
 }
+
+class MultipleStitchExpanderTest: XCTestCase {
+
+
+    func testNestedArrayBuilder() throws {
+        let result = NestedArrayBuilder().expandRow(row: ["k4", "p1"])
+        expect(result).to(equal(["k1", "k1", "k1", "k1", "p1"]))
+    }
```
Comment (hnlee) [inline Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift:26]: Let's test some edge cases! What happens when you input "kx py"?

Comment (hnlee) [reply Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift:26]: As you said in pairing, let's also cover "k0"!

**Review by hnlee — CHANGES_REQUESTED** [review]

**Comment by hnlee** [comment]
Other ideas from pairing:
- Make `StitchLibrary` its own class.
- Think about... do we need a `Row` object (not just the metadata about the row, but an abstraction to represent the collection of stitches that is the row itself)?

File: Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift (hunk 3 above)
Comment (graceolivia) [reply Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift:26]: Added these!

File: Sources/SwiftCliCore/NestedArrayBuilder.swift (hunk 1 above)
Comment (graceolivia) [reply Sources/SwiftCliCore/NestedArrayBuilder.swift:14]: Done!

File: Sources/SwiftCliCore/InputValidator.swift (hunk 2 above)
Comment (graceolivia) [reply Sources/SwiftCliCore/InputValidator.swift:35]: Changed the name and abstracted out the error handling! It feels more readable now. I also considered bundling up the error handling with the data transformation that happens with it at the same time (so a function would take `[String]`, run the `EmptyRowError` check, and then transform to `[[String]]`), but that seems to invalidate the single responsibility rule.  What do you think?

**Comment by graceolivia** [comment]
> Other ideas from pairing:
> 
> * Make `StitchLibrary` its own class.
> * Think about... do we need a `Row` object (not just the metadata about the row, but an abstraction to represent the collection of stitches that is the row itself)?

Not from this commit: If it's okay with you, I'm holding off on these for now because I need to resubmit my other PRs first, but I'll circle back once I have time.

File: Charts/.gitignore [hunk 4]
```diff
@@ -0,0 +1,3 @@
+*
+*/
+!.gitignore
```
Comment (hnlee) [inline Charts/.gitignore:3]: Hm, I would add these to the  `.gitignore` in the project root.

File: Sources/SwiftCliCore/InputValidator.swift [hunk 5]
```diff
@@ -3,8 +3,26 @@ import Foundation
...
+            patternNestedArray = knitFlatArray(array: patternNestedArray)
+        }
+
+        try checkNoInvalidStitchesInNestedArray(pattern: patternNestedArray)
+
+        let expandedNestedArray = try patternNestedArray.map { try NestedArrayBuilder().expandRow(row: $0) }
+
+        let patternMetaData = MetaDataBuilder().buildAllMetaData(stitchArray: expandedNestedArray)
+
+        return try checkNoMathmaticalIssuesInArrayOfRowInfo(pattern: patternMetaData)
```
Comment (hnlee) [inline Sources/SwiftCliCore/InputValidator.swift:21]: Typo here.

File: Sources/SwiftCliCore/InputValidator.swift [hunk 6]
```diff
@@ -29,21 +45,35 @@ public class InputValidator {
...
-            return patternMetaData
+            return pattern
         case .failure(let error):
             throw error
         }
     }
 
 
+
+    private func check(pattern: [[String]]) throws -> [[String]] {
```
Comment (hnlee) [inline Sources/SwiftCliCore/InputValidator.swift:63]: Is this a duplicate of `checkNoInvalidStitchesInNestedArray()`?

**Review by hnlee — CHANGES_REQUESTED** [review]
Just some clean up to do.

File: Charts/.gitignore (hunk 4 above)
Comment (graceolivia) [reply Charts/.gitignore:3]: Okay! I needed to put something in there since otherwise git will ignore the directory, maybe I'll put a brief README instead.

Comment (breadoliveoilsalt) [reply Charts/.gitignore:3]: Grace, holler at me if you want to talk about this or have any questions.  (I'm a bit confused by git's behavior here.)

File: Sources/SwiftCliCore/Chartify.swift [hunk 7]
```diff
@@ -40,7 +40,7 @@ public final class Chartify {
     }
 
     private func validateAndChartify(pattern: [String], knitFlat: Bool) throws -> String {
-        let metaData = try inputValidator.inputValidation(pattern: pattern, knitFlat: knitFlat)
+        let metaData = try inputValidator.validateInput(pattern: pattern, knitFlat: knitFlat)
```
Comment (breadoliveoilsalt) [inline Sources/SwiftCliCore/Chartify.swift:43]: Nice rename.

File: Sources/SwiftCliCore/InputValidator.swift (hunk 6 above)
Comment (graceolivia) [reply Sources/SwiftCliCore/InputValidator.swift:63]: Oops! Thank you, deleted.

File: Sources/SwiftCliCore/NestedArrayBuilder.swift [hunk 8]
```diff
@@ -8,4 +8,35 @@ public class NestedArrayBuilder {
...
+    }
+
+    private func getMultipleStitchInfo(stitch: String) throws -> [String] {
+        let isRepeatingStitch = repeatingStitches.first(where: { stitch.starts(with: $0.name )})
+        if let isRepeatingStitch = isRepeatingStitch {
+            let stitchName = isRepeatingStitch.name
+            var clipStitch = stitch
+            clipStitch.removeAll(where: { stitchName.contains($0) })
+            let stitchNameSuffix = clipStitch
+            guard let repeatNumber = Int(stitchNameSuffix) else {
```
Comment (breadoliveoilsalt) [inline Sources/SwiftCliCore/NestedArrayBuilder.swift:23]: Not a blocker -- something for pairing later after you merge -- let's talk about refactoring the names here and maybe making a little clearer.

File: Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift [hunk 9]
```diff
@@ -16,3 +16,23 @@ class NestedArrayMakerTest: XCTestCase {
...
+
+    func testInvalidZeroCountStitch() throws {
+
+        expect{ try NestedArrayBuilder().expandRow(row: ["k0", "p1"]) }.to(throwError(InputError.invalidStitchNumber(rowNumber: nil, invalidStitch: "k0", validStitchType: "k", invalidStitchNumber: "0")))
+    }
+
+    func testInvalidWordCountStitch() throws {
+
+        expect{ try NestedArrayBuilder().expandRow(row: ["kx", "py"]) }.to(throwError(InputError.invalidStitchNumber(rowNumber: nil, invalidStitch: "kx", validStitchType: "k", invalidStitchNumber: "x")))
+    }
```
Comment (breadoliveoilsalt) [inline Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift:36]: You can delete the extra lines above the expect.  Also, maybe play with leaving the code as is, but seeing if swift will allow you to break it into separate lines for readibility.  Like

`var obj = {name: "Tony", age: "Go away"}`
vs
```
var obj = {
  name: "Tony", 
  age: "Go away"
}
```

File: Charts/.gitignore (hunk 4 above)
Comment (hnlee) [reply Charts/.gitignore]: FWIW, the `Charts/` directory ought to be created by the program if it doesn't exist, so I would actually just have `git` ignore the directory too.

**Comment by breadoliveoilsalt** [comment]
As of 10:30am Friday 11/5 -- expect no other comments from me!

File: Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift (hunk 9 above)
Comment (graceolivia) [reply Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift:36]: That is more readable!
